src/software/ai/navigator/obstacle/obstacle.h

```diff
@@ -1,14 +1,23 @@
 #pragma once
 
 #include <sstream>
+#include <variant>
 
 #include "shared/constants.h"
+#include "software/new_geom/circle.h"
+#include "software/new_geom/convex_polygon.h"
 #include "software/new_geom/point.h"
 #include "software/new_geom/polygon.h"
 #include "software/new_geom/segment.h"
 #include "software/new_geom/util/distance.h"
 #include "software/new_geom/util/intersects.h"
 
+/**
+ * We use this variant when we need to know the underlying shape of the obstacle, e.g. in
+ * tests and for drawing with the visualizer
+ */
+using ObstacleShape = std::variant<ConvexPolygon, Circle>;
```

Comment:
We shouldn't need to do this if we have a visitor?

Reply:
this is needed because we are using `std::visit`, not our own visitor class structure.

Reply:
Ah, if we're going to go with a visitor (might not need to if it was only for tests), then we should properly implement the visitor pattern like we have elsewhere (ex. for `Primitive`).

Reply:
the overloaded lambda visitor only works on variants so I have ObstaclePtr->getObstacleShape and operate std::visit on that

Reply:
Hmm, it's probably worth implementing the visitor design pattern properly for all our geometry types, otherwise every time we add a new geometry type as an obstacle we're going to need to go through and manually update all the `std::visit` calls.